tests/package/test_packagers_manager.py

```diff
@@ -0,0 +1,364 @@
+# Copyright 2018 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import os
+import shutil
+import tempfile
+import zipfile
+from typing import Any, Dict, List, Tuple, Type, Union
+
+import pytest
+
+from mlrun import DataItem
+from mlrun.artifacts import Artifact
+from mlrun.errors import MLRunInvalidArgumentError
+from mlrun.package import (
+    DefaultPackager,
+    MLRunPackagePackagerCollectionError,
+    Packager,
+    PackagersManager,
+)
+
+
+class PackagerA(Packager):
+    """
+    A simple packager to pack strings as results.
+    """
+
+    PACKABLE_OBJECT_TYPE = str
+
+    @classmethod
+    def get_default_artifact_type(cls, obj: Any) -> str:
+        return "result"
+
+    @classmethod
+    def get_supported_artifact_types(cls) -> List[str]:
+        return ["result"]
+
+    @classmethod
+    def is_packable(cls, object_type: Type, artifact_type: str = None) -> bool:
+        return object_type is cls.PACKABLE_OBJECT_TYPE and artifact_type == "result"
+
+    @classmethod
+    def pack(
+        cls, obj: str, artifact_type: Union[str, None], configurations: dict
+    ) -> dict:
+        return {f"{configurations['key']}_from_PackagerA": obj}
+
+    @classmethod
+    def unpack(
+        cls,
+        data_item: DataItem,
+        artifact_type: Union[str, None],
+        instructions: dict,
+    ) -> str:
+        pass
+
+
+class PackagerB(DefaultPackager):
```

Comment:
At the end of the day, I'm a bit confused as to when I should inherit from `Packager` and when I need to extend `DefaultPackager`. Should I extend `DefaultPackager` when I want to support all types and implement a new packaging format? Or is there another reasoning?

Reply:
Default packager gives you quality of life features like already implementing logging as result and object (pickling) and also implements for you automatically all the abstract methods. Preferably a user would like to inherit from the `DefaultPackager`, but we don't force him to. He can start from scratch by inheriting `Packager`.